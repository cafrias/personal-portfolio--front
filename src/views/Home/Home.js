// @flow
import React, {Component} from 'react'
import {Motion, spring, presets} from 'react-motion'

import './Home.sass'

// COMPONENTS __________________________________________________________________
import Arrow from '../../components/Arrow'

// ANIMATIONS __________________________________________________________________
import FadeIn from '../../animations/FadeIn'

// Define final text
const FIRST_LINE = 'I am'
const NAME = 'Carlos Frias'
const SECOND_LINE = 'full-stack'
const THIRD_LINE = 'ux/ui'

// TYPES _______________________________________________________________________
type HomeProps = {
  setNextPage: (string) => void,
  inTransitionAnimation: [boolean, string]
}

type HomeState = {
  firstLine: string,
  name: string,
  secondLine: string,
  thirdLine: string,
  finishedTyping: boolean,
  startedTyping: boolean,
  developer: boolean,
  designer: boolean,
  ready: boolean
}

class Home extends Component {
  props: HomeProps
  state: HomeState

  firstLine: Array<string>
  secondLine: Array<string>
  name: Array<string>
  thirdLine: Array<string>
  typeCharArray: (Array<string>, string) => Promise<any>
  displaySub: (string) => Promise<any>

  constructor(props: HomeProps) {
    super(props)

    this.firstLine = FIRST_LINE.split('')
    this.name = NAME.split('')
    this.secondLine = SECOND_LINE.split('')
    this.thirdLine = THIRD_LINE.split('')

    this.state = {
      firstLine: '',
      name: '',
      secondLine: '',
      thirdLine: '',
      finishedTyping: false,
      startedTyping: false,
      developer: false,
      designer: false,
      ready: false
    }

    this.typeCharArray = this.typeCharArray.bind(this)
    this.displaySub = this.displaySub.bind(this)
  }

  startTyping() {
    const {
      typeCharArray,
      displaySub,
      firstLine,
      name,
      secondLine,
      thirdLine
    } = this

    // Type `.name`
    typeCharArray(firstLine, 'firstLine')
      // Type `.semi`
      .then(() => typeCharArray(name, 'name'))
      // Type 'FULL-STACK'
      .then(() => typeCharArray(secondLine, 'secondLine'))
      // Trigger 'developer' animation
      .then(() => displaySub('developer'))
      // Type 'UX/UI'
      .then(() => typeCharArray(thirdLine, 'thirdLine'))
      // Trigger 'designer' animation
      .then(() => displaySub('designer'))
      // Finish typing
      .then(() => this.setState({finishedTyping: true}))
  }

  displaySub(property: string): Promise<any> {
    return new Promise((resolve, reject) => {
      this.setState({
        [property]: true
      })
      setTimeout(() => resolve(), 800)
    })
  }

  typeCharArray(charArray: Array<string>, property: string): Promise<any> {
    return new Promise((resolve) => {
      const length = charArray.length

      charArray.forEach((char, i) => {
        const final = (i + 1) === length
        setTimeout(() => requestAnimationFrame(() => {
          this.setState({
            [property]: this.state[property] + char,
            startedTyping: true
          })
          // Resolve if finished
          if(final) resolve(true)
        }), (i + 1) * 120)
      })
    })
  }

  componentDidUpdate() {
    const {ready, startedTyping} = this.state
    if(ready && !startedTyping) this.startTyping()
  }

  titleWithSub(title: string, subtitle: string, shorter: boolean, first: boolean = false) {
    return (
      <div className={`title-sub ${shorter ? 'title-sub--half' : ''}`}>
        {
          first
          ? (
            <p className="name">
              {this.state.firstLine}
              <span className="semi">{' ' + this.state.name}</span>
            </p>
          )
          : null
        }
        <p className="job-title">{title}</p>
        {
          this.state[subtitle]
          ? (
            <Motion
              defaultStyle={{s: 0.6, o: 0}}
              style={{s: spring(1, presets.wobbly), o: spring(1)}}
            >
              {
                ({s, o}) => (
                  <span style={{
                      opacity: o,
                      transform: `scale(${s}, ${s})`
                    }} className={`job-sub ${shorter ? 'job-sub--shorter' : ''}`}>
                    {subtitle}
                  </span>
                )
              }
            </Motion>
          )
          : null
        }
      </div>
    )
  }

  mountChildren() {
    const {
      secondLine,
      thirdLine,
      finishedTyping
    } = this.state

    // TODO: Check if Home is read by screen reader
    return (
        <div className={`home-content ${finishedTyping ? 'home-content--pic' : ''}`}>
          <h1 id="home__title" hidden>Home</h1>
          <div className="presentation">
            {
              this.titleWithSub(secondLine, 'developer', false, true)
            }
            {
              this.titleWithSub(thirdLine, 'designer', true)
            }
          </div>
          <div className="controls">
            {
              finishedTyping
              ? (<Arrow link="/about" />)
              : null
            }
          </div>
        </div>
    )
  }

  componentDidMount() {
    this.props.setNextPage('/about')
  }

  render() {
    const {ready} = this.state
    const {inTransitionAnimation} = this.props
    return(
      <FadeIn
        onRest={() => {this.setState({ready: true})}}
      >
        <section id="home" className={`home-container ${
            inTransitionAnimation[0]
              ? 'home-container--transition'
              : ''
          }`}>
          <div className="curtain">
            {
              ready
              ? this.mountChildren()
              : null
            }
          </div>
        </section>
      </FadeIn>
    )
  }
}

export default Home
