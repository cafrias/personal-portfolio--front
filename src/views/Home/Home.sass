@import '../../_sass/_constants.sass'
@import '../../_sass/_breakpoints.sass'

.home-container
  background-image: url('./hero.png')
  background-repeat: no-repeat
  background-position: left bottom
  background-size: cover
  border-top: 6.854rem transparent solid
  width: 100%
  height: 100vh
  overflow: hidden

.home-content
  padding-top: 10%
  width: 100%
  height: 100%
  display: flex
  flex-direction: column
  justify-content: space-between
  align-items: center
  +anti-aliased

.curtain
  width: 100%
  height: 100%
  background-color: rgba(0, 0, 0, .75)


// PRESENTATION ________________________________________________________________
.presentation
  line-height: 1

.name
  font-size: 1.618rem
  color: $primary
  margin: 0
  margin-bottom: 1.618rem
  align-self: flex-start

.semi
  font-weight: bold
  color: $white
  text-transform: uppercase

// TITLE WITH SUB ______________________________________________________________
.title-sub
  color: $white
  display: flex
  flex-direction: column
  align-items: flex-end
  margin-bottom: 2.618rem

.job-title
  font-size: 3.272rem
  font-weight: 900
  margin-top: 0
  margin-bottom: 0.773rem
  text-transform: uppercase

.job-sub
  font-size: 1.618rem

.job-sub::before
  content: ''
  height: .5rem
  width: 8.567rem
  display: inline-block
  border-top: 1px $primary solid
  margin-right: 1rem

.job-sub--shorter::before
  width: 1.618rem

// CONTROLS ____________________________________________________________________
.controls
  flex-grow: 2
  display: flex
  flex-direction: column
  justify-content: flex-end

// LANDSCAPE BREAKPOINT -- MOBILE ______________________________________________
+for-phone-portrait
  .home-container
    border-top: 3.272rem rgba(0, 0, 0, .75) solid
    background-position: center

  .presentation
    display: flex
    justify-content: space-around
    align-items: flex-end
    width: 100%

  .home-content
    padding-top: 1.618rem

// TABLET BREAKPOINT -- PORTRAIT _______________________________________________
+for-tablet-portrait-up
  .home-container
    background-position: center

  .home-content
    padding-top: 20%

  // PRESENTATION
  .presentation
    padding-right: 3.272rem
    align-self: flex-end

  // TITLE WITH SUB
  .name
    font-size: 1.618rem

  .job-title
    font-size: 5.295rem

  .job-sub
    font-size: 2.023rem

  .job-sub::before
    width: 17.942rem

  .job-sub--shorter::before
    width: 5.295rem

  // CONTROLS
  .controls
    width: 50%
    align-items: center

// TABLET BREAKPOINT -- PORTRAIT _______________________________________________
+for-tablet-landscape-up
  .home-content
    padding-top: 5%

  .curtain
    background-color: transparent

  .home-container
    background-position: left bottom
    background-size: contain

// DESKTOP BREAKPOINT __________________________________________________________
+for-desktop-up
  .home-container
    border-top: 4.236rem transparent solid

  .home-content
    padding-top: 10%
    width: 90%
    align-items: flex-end

  .presentation
    padding: 0
    width: 46%
    display: flex
    flex-direction: column
    align-items: flex-end

  .name
    font-size: 2.023rem

  // TITLE WITH SUB
  .title-sub
    width: 100%

  .title-sub--half
    width: 75%

  .job-title
    font-size: 5.295rem
    align-self: flex-start

  .job-sub
    font-size: 2.618rem

  .job-sub--shorter::before
    width: 11.089rem

// DESKTOP BIG BREAKPOINT ______________________________________________________
+for-big-desktop-up
  .job-title
    font-size: 6.854rem

  .job-sub
    font-size: 3.272rem

  .name
    font-size: 2.618rem

  .job-sub::before
    width: 22.428rem

  .job-sub--shorter::before
    width: 13.861rem
