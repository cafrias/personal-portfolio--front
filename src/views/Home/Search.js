import React from 'react'

import './Search.sass'
import glass from './mag-glass.svg'

function Search() {
  return (
    <div className="search">
      <div className="mag-glass-container">
        <img src={glass} alt="Search" className="mag-glass"/>
      </div>
    </div>
  )
}

export default Search
