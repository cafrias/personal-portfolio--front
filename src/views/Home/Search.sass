@import '../../_sass/_breakpoints.sass'

// SEARCH ______________________________________________________________________
.search
  flex-grow: 2
  display: flex
  flex-direction: column
  align-items: center

// MAG GLASS ___________________________________________________________________
.mag-glass-container
  display: inline-block
  border: 2px solid $dark-grey
  padding: 0.5rem 0.6rem
  border-radius: 50%

// TABLET BREAKPOINT -- PORTRAIT _______________________________________________
+for-tablet-portrait-up
  .search
    justify-content: center
