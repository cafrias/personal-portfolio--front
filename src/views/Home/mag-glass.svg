<svg xmlns="http://www.w3.org/2000/svg" width="25" height="25" viewBox="0 0 30 30">
  <g fill="#9b9b9b">
    <path d="M26.0502294,3.3315 C21.7593818,-1.1106 14.803091,-1.1106 10.5122434,3.3315 C6.2213958,7.7733 6.2213958,14.9751 10.5122434,19.4172 C14.803091,23.8593 21.7593818,23.8593 26.0502294,19.4172 C30.341077,14.9751 30.341077,7.7736 26.0502294,3.3315 Z M24.8235692,18.1476 C21.2102391,21.888 15.3522338,21.888 11.7389036,18.1476 C8.12586332,14.4069 8.12586332,8.3421 11.7389036,4.6017 C15.3522338,0.8613 21.2105289,0.861 24.8235692,4.6017 C28.4366095,8.3424 28.4366095,14.4066 24.8235692,18.1476 Z"/>
    <rect width="3.47" height="11.973" x="3.87" y="18.51" transform="rotate(45 5.605 24.496)"/>
  </g>
</svg>
