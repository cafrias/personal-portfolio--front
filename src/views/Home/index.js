import React, {Component} from 'react'
import {Motion, spring, presets} from 'react-motion'

import './Home.sass'
import glass from './mag-glass.svg'

// COMPONENTS __________________________________________________________________=
import Arrow from '../../components/Arrow'

// Define final text
const FIRST_LINE = 'I am'
const NAME = 'Carlos Frias'
const SECOND_LINE = 'full-stack'
const THIRD_LINE = 'ux/ui'

class Home extends Component {
  constructor(props) {
    super(props)

    this.firstLine = FIRST_LINE.split('')
    this.name = NAME.split('')
    this.secondLine = SECOND_LINE.split('')
    this.thirdLine = THIRD_LINE.split('')

    this.state = {
      firstLine: '',
      name: '',
      secondLine: '',
      thirdLine: '',
      finishedTyping: false,
      developer: false,
      designer: false
    }

    this.typeCharArray = this.typeCharArray.bind(this)
    this.displaySub = this.displaySub.bind(this)
  }

  componentDidMount() {
    this.startTyping()
  }

  startTyping() {
    const {
      typeCharArray,
      displaySub,
      firstLine,
      name,
      secondLine,
      thirdLine
    } = this

    // Type `.name`
    typeCharArray(firstLine, 'firstLine')
      // Type `.semi`
      .then(() => typeCharArray(name, 'name'))
      // Type 'FULL-STACK'
      .then(() => typeCharArray(secondLine, 'secondLine'))
      // Trigger 'developer' animation
      .then(() => displaySub('developer'))
      // Type 'UX/UI'
      .then(() => typeCharArray(thirdLine, 'thirdLine'))
      // Trigger 'designer' animation
      .then(() => displaySub('designer'))
      // Finish typing
      .then(() => this.setState({finishedTyping: true}))
  }

  displaySub(property) {
    return new Promise((resolve, reject) => {
      this.setState({
        [property]: true
      })
      setTimeout(() => resolve(), 800)
    })
  }

  typeCharArray(charArray, property) {
    return new Promise((resolve) => {
      const length = charArray.length

      charArray.forEach((char, i) => {
        const final = (i + 1) === length
        setTimeout(() => requestAnimationFrame(() => {
          this.setState({
            [property]: this.state[property] + char,
          })
          // Resolve if finished
          if(final) resolve(true)
        }), (i + 1) * 120)
      })
    })
  }

  titleWithSub(title, subtitle, shorter, first) {
    return (
      <div className={`title-sub ${shorter ? 'title-sub--half' : ''}`}>
        {
          first
          ? (
            <p className="name">
              {this.state.firstLine}
              <span className="semi">{' ' + this.state.name}</span>
            </p>
          )
          : null
        }
        <p className="job-title">{title}</p>
        {
          this.state[subtitle]
          ? (
            <Motion
              defaultStyle={{s: 0.6, o: 0}}
              style={{s: spring(1, presets.wobbly), o: spring(1)}}
            >
              {
                ({s, o}) => (
                  <span style={{
                      opacity: o,
                      transform: `scale(${s}, ${s})`
                    }} className={`job-sub ${shorter ? 'job-sub--shorter' : ''}`}>
                    {subtitle}
                  </span>
                )
              }
            </Motion>
          )
          : null
        }
      </div>
    )
  }

  render() {
    const {
      secondLine,
      thirdLine,
      finishedTyping
    } = this.state
    // TODO: Check if Home is read by screen reader
    return (
      <section id="home" className="home-container">
        <div className={`home-content ${finishedTyping ? 'home-content--pic' : ''}`}>
          <h1 id="home__title" hidden>Home</h1>
          <div className="presentation">
            {
              this.titleWithSub(secondLine, 'developer', false, true)
            }
            {
              this.titleWithSub(thirdLine, 'designer', true)
            }
          </div>
          <div className="controls">
            <div className="search">
              <div className="mag-glass-container">
                <img src={glass} alt="Search" className="mag-glass"/>
              </div>
            </div>
            {
              finishedTyping
              ? (<Arrow />)
              : null
            }
          </div>
        </div>
      </section>
    )
  }
}

export default Home
