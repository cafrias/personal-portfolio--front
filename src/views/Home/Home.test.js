import React from 'react'
import {shallow} from 'enzyme'

import Home from './index'

describe('Home', () => {
  const homeWrapper = shallow(<Home />)

  it('should render without crashing', () => {
    expect(homeWrapper).toHaveLength(1)
  })

  it('should match snapshot', () => {
    expect(homeWrapper).toMatchSnapshot()
  })
})
