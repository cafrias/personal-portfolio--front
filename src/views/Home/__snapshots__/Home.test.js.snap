exports[`Home should match snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <FadeIn
    onRest={[Function]}>
    <section
        className="home-container"
        id="home">
        <div
            className="curtain" />
    </section>
</FadeIn>,
  "nodes": Array [
    <FadeIn
      onRest={[Function]}>
      <section
            className="home-container"
            id="home">
            <div
                  className="curtain" />
      </section>
</FadeIn>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Home />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Home {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "displaySub": [Function],
        "firstLine": Array [
          "I",
          " ",
          "a",
          "m",
        ],
        "name": Array [
          "C",
          "a",
          "r",
          "l",
          "o",
          "s",
          " ",
          "F",
          "r",
          "i",
          "a",
          "s",
        ],
        "props": Object {},
        "refs": Object {},
        "secondLine": Array [
          "f",
          "u",
          "l",
          "l",
          "-",
          "s",
          "t",
          "a",
          "c",
          "k",
        ],
        "state": Object {
          "designer": false,
          "developer": false,
          "finishedTyping": false,
          "firstLine": "",
          "name": "",
          "ready": false,
          "secondLine": "",
          "startedTyping": false,
          "thirdLine": "",
        },
        "thirdLine": Array [
          "u",
          "x",
          "/",
          "u",
          "i",
        ],
        "typeCharArray": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <FadeIn
          onRest={[Function]}>
          <section
                    className="home-container"
                    id="home">
                    <div
                              className="curtain" />
          </section>
</FadeIn>,
        "_debugID": 2,
        "_renderedOutput": <FadeIn
          onRest={[Function]}>
          <section
                    className="home-container"
                    id="home">
                    <div
                              className="curtain" />
          </section>
</FadeIn>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Home />,
}
`;
